### ISSUE TITLE
Unable to use MSI with AKS on Azure

### ISSUE THREAD

**Issue by garrypolley** [Azure/AKS#218]
I've logged a [stackoverflow](https://stackoverflow.com/questions/49061349/unable-to-use-msi-with-aks-on-azure) post about this issue. I'll copy it here for ease of use: 

I have a k8s cluster in Azure using the AKS preview.

I also have MSI enabled on the VMs. If I ssh to the VMs I can see the MSI service is working:

```sh
curl http://localhost:50342/oauth2/token --data "resource=https://vault.azure.net" -H Metadata:true
```

I get a response that has what you'd expect based on the documentation.

However, I cannot figure out how to get my aspnet core applications to use the AzureServiceTokenProvider as desired. Locally it works, once deployed inside a pod in kubernetes it cannot find the authorization required.

The error I get is:

> Microsoft.Azure.Services.AppAuthentication.AzureServiceTokenProviderException: Parameters: Connectionstring: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/9d93c307-6856-4bab-8fa9-99690e0fabaf. Exception Message: Tried the following 3 methods to get an access token, but none of them worked. Parameters: Connectionstring: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/9d93c307-6856-4bab-8fa9-99690e0fabaf. Exception Message: Tried to get token using Managed Service Identity. Unable to connect to the Managed Service Identity (MSI) endpoint. Please check that you are running on an Azure resource that has MSI setup. Parameters: Connectionstring: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/9d93c307-6856-4bab-8fa9-99690e0fabaf. Exception Message: Tried to get token using Visual Studio. Access token could not be acquired. Environment variable LOCALAPPDATA not set. Parameters: Connectionstring: [No connection string specified], Resource: https://vault.azure.net, Authority: https://login.windows.net/9d93c307-6856-4bab-8fa9-99690e0fabaf. Exception Message: Tried to get token using Azure CLI. Access token could not be acquired. /bin/bash: az: No such file or directory 

Has anyone been able to read from KeyVault in a pod using similar code?

```C#
AzureServiceTokenProvider azureServiceTokenProvider = new 
AzureServiceTokenProvider();
KVC = new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(azureServiceTokenProvider.KeyVaultTokenCallback));
var baseUrl = config.GetValue<string>("Azure:VaultURL");
var secretLocation = config.GetValue<string>("Azure:SecretLocation");
location = $"{baseUrl}{secretLocation}";
KVC.GetSecretAsync(location).Result.Value;
```

Yes. I’ve set up a Daemon set with host networking to have access to the VM localhost. What I cannot figure out is how to get the MSI_SECRET and MSI_ENDPOINT set. My plan is to proxy to the Daemon set for the pods on the node. SSH to the node and I do not see those environment vars set. But I can hit the oauth service.

I’m wanting to avoid hard coding secrets. My understanding is the purpose of MSI is to remove the need for hard coded secrets. My last resort will be hard coding and storing those values.

I appreciate any help with this issue. If the answer is to hard code my keys and secrets for now I can accept that.

**Comment by andyzhangx** [comment]
MSI is for a host VM, while your request is for a pod, AFAIK, it does work that way.
 another workaround way is use configmap to store secret, thus it could be configured,  see example here:
https://github.com/andyzhangx/demo/tree/master/windows/configmap

**Comment by garrypolley** [comment]
> MSI is for a host VM, while your request is for a pod, AFAIK, it does work that way

My desire is to use the MSI service for getting the tokens in a way that is more secure. My understanding of MSI is that it would allow an application running on the VM to get credentials in a more secure manner, than hard coding them.

> another workaround way is use configmap to store secret

More or less, it would allow me to _not_ have to store keys and secrets in a config map or a k8s secret. I want to avoid storing credentials in any form of source control.  

> MSI is for a host VM

I experimented with using a host networked Daemon Set to proxy to the localhost VM, however, I couldn't get access to the `MSI_SECRET` or the `MSI_ENDPOINT` even if I SSH'd onto the machine. After hard coding the `MSI_ENDPOINT` to the localhost and port that seemed to work (for hitting the service. However, I still couldn't get the `MSI_SECRET` so my experiment failed. 😞 

It worked for pod-to-pod communication. Without the `MSI_SECRET` the tokens "granted" did not appear to actually have access to any resources.  


My approach for now is to use a secret in k8s that is namespaced to my app. 

Are there plans to allow something akin to the MSI service to be usable within k8s with AKS?

**Comment by bradwilson** [comment]
Any word on when MSI will be usable from inside of AKS containers?

**Comment by colemickens** [comment]
Since it's not linked here... https://github.com/Azure/aad-pod-identity

**Comment by Pinox** [comment]
MSI should be working in AKS pods like in App Services. What the point of having key vault in Azure when it does not work also in AKS pods.

Same applies to Reserved IP addresses , should also work in AKS Load balancer , current solution to use Static IP is to create one on AKS  cluster. Note - generated one not one used from Reserved IP's

So Reserved IP's and MSI are currently not working as intended in AKS pods.

**Comment by ritazh** [comment]
To access keyvault content from a pod, take a look at https://github.com/Azure/kubernetes-keyvault-flexvol This mounts content from keyvault into your pod and works with pod identity.

**Comment by seanmck** [comment]
Closing because the pod identity and keyvault flexvol projects should suffice for this. Note that intend to integrate these directly into AKS in the coming months.

**Comment by erick-thompson** [comment]
Did these get merged into AKS?

**Comment by ryanmaclean** [comment]
@seanmck - where did this leave off?

**Comment by sirdank** [comment]
MSI auth to key vault from inside the pod is working as expected for me.

**Comment by lpghartz** [comment]
For the benefit of anyone who runs into this, as it took me a bit of flailing to figure this out. If your goal of having MSI_ENDPOINT / MSI_SECRET is to be able to deploy existing code using the Azure SDKs that rely on them for MSI authentication on AppService, azure-pod-identity is compatible with the MSI proxy running in AppService, you just need to inject the environment variables

MSI_ENDPOINT is fixed: http://169.254.169.254/metadata/identity/oauth2/token/
MSI_SECRET isn't used, but the SDKs need it, so set it to anything.

This in an env block of your deployment "just works":
`
env:
        - name: MSI_ENDPOINT
          value: http://169.254.169.254/metadata/identity/oauth2/token/
        - name: MSI_SECRET
          value: ""
`

It'll correctly pick up the identity associated with the "aadpodidbinding" and return the token to the Azure SDK.
